diff -urNwB linux-2.6.25.8/arch/arm/plat-s3c24xx/common-smdk.c linux-2.6.25.8_EmbedSky/arch/arm/plat-s3c24xx/common-smdk.c
--- linux-2.6.25.8/arch/arm/plat-s3c24xx/common-smdk.c	2008-06-22 13:25:26.000000000 +0800
+++ linux-2.6.25.8_EmbedSky/arch/arm/plat-s3c24xx/common-smdk.c	2009-02-28 13:27:40.000000000 +0800
@@ -108,44 +108,19 @@
 
 static struct mtd_partition smdk_default_nand_part[] = {
 	[0] = {
-		.name	= "Boot Agent",
-		.size	= SZ_16K,
-		.offset	= 0,
+		.name	= "TQ2440_uboot",
+		.size	= 0x00040000,
+		.offset	= 0x00000000,
 	},
 	[1] = {
-		.name	= "S3C2410 flash partition 1",
-		.offset = 0,
-		.size	= SZ_2M,
+		.name	= "TQ2440_kernel",
+		.offset	= 0x0004C000,
+		.size	= 0x00200000,
 	},
 	[2] = {
-		.name	= "S3C2410 flash partition 2",
-		.offset = SZ_4M,
-		.size	= SZ_4M,
-	},
-	[3] = {
-		.name	= "S3C2410 flash partition 3",
-		.offset	= SZ_8M,
-		.size	= SZ_2M,
-	},
-	[4] = {
-		.name	= "S3C2410 flash partition 4",
-		.offset = SZ_1M * 10,
-		.size	= SZ_4M,
-	},
-	[5] = {
-		.name	= "S3C2410 flash partition 5",
-		.offset	= SZ_1M * 14,
-		.size	= SZ_1M * 10,
-	},
-	[6] = {
-		.name	= "S3C2410 flash partition 6",
-		.offset	= SZ_1M * 24,
-		.size	= SZ_1M * 24,
-	},
-	[7] = {
-		.name	= "S3C2410 flash partition 7",
-		.offset = SZ_1M * 48,
-		.size	= SZ_16M,
+		.name	= "TQ2440_yaffs2",
+		.offset	= 0x0024C000,
+		.size	= 0x03DB0000,
 	}
 };
 
@@ -163,9 +138,9 @@
 */
 
 static struct s3c2410_platform_nand smdk_nand_info = {
-	.tacls		= 20,
-	.twrph0		= 60,
-	.twrph1		= 20,
+	.tacls		= 10,
+	.twrph0		= 25,
+	.twrph1		= 10,
 	.nr_sets	= ARRAY_SIZE(smdk_nand_sets),
 	.sets		= smdk_nand_sets,
 };
